Add `tape.toml` for configs

aarjn opened the pull request:
PR for #16 

I have pushed code of simple implementation of using `tape.toml` . In the following code change of only `rpc_url` is used from `tape.toml` config file. Couple of changes I made

- add `toml` crate for handling toml file
- removed default cluster option of `l` from cli args
- priority is given to cli commands. if cluster is mention in cli command, `rpc_url` from `tape.toml` wont be used. if not it will be used. 
- if config is loaded from `tape.toml` it is passed through cli context 
- `tape.toml` is created at home dir 

I tested the above change in `archive` command

`~/tape.toml`
```toml
[solana]
rpc_url = "https://api.devnet.solana.com"
ws_url = "wss://api.devnet.solana.com/"
commitment = "confirmed"
```

- `cargo run -p tapedrive-cli --manifest-path ~/Documents/solana/tape/cli/Cargo.toml archive`
   <img width="463" height="91" alt="Screenshot 2025-08-07 at 1 35 20 PM" src="https://github.com/user-attachments/assets/57811755-3dd7-4812-83b7-5bda15a2f1e4" />

- `cargo run -p tapedrive-cli --manifest-path ~/Documents/solana/tape/cli/Cargo.toml archive -ul`
   <img width="374" height="90" alt="Screenshot 2025-08-07 at 1 35 53 PM" src="https://github.com/user-attachments/assets/31c077a3-2327-4760-a6d9-e59bdcc95630" />

Nagaprasadvr commented:
Cool ,
can u disable rust formatting, this repo is not using rust formatting and the pr will add unnecessary changes

zfedoran commented:
Welcome @4rjunc, I hope this is one of many PRs, I'd love to see more in the future.

I'm curious, why are we removing the local cluster option?

4rjunc commented:
> Also i feel we dont wanna use cli args because the reason for adding toml is to make things easier to configure wdyt @zfedoran

Quoting this again in case you missed replying @zfedoran

aarjn committed:
feat: update fields for TapeCofig

> transaction
> performance
> identity
> solana
> storage
> network
> logging

feat: add validate + default values for toml.config

feat: log_level is enum

fix: resolving comments

> change the `max_memory_mb` 2gb -> 16 gb
> add few comments on what to change

fix: remove the `NetworkConfig`

feat: add enum for commitment level

> add a impl for `CommitmentLevel`

fix: fix  the keypair loading up from ~ path

feat: removed the helper function for tilde, add function used in anchor

feat: add abstract StorageConfig

feat: add URL validation

fix: rename performance -> mining_config

> add max_poa_thread
> add max_pow_thread

feat: better error handling for config::load()

- ConfigFileNotFound
- InvalidUrl(String)
- KeypairNotFound(String)
- HomeDirectoryNotFound
- FileReadError(std::io::Error)
- ParseError(toml::de::Error)
- DefaultConfigCreationFailed(String)

feat: add `--config` option

feat: add error handling for config load

> create new config is not found in default path, `--config` is not used
> if config file not found in given path -> error will be shown

feat: add `--config` option

feat: add error handling for config load

> create new config is not found in default path, `--config` is not used
> if config file not found in given path -> error will be shown

Nagaprasadvr reviewed (changes requested):
few things to update , can u also add a unit test to test the toml parsing works properly

4rjunc commented on cli/src/config.rs:
```diff
@@ -0,0 +1,148 @@
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TransactionConfig {
+    pub priority_fee: u64,
```
updated

Nagaprasadvr commented:
> > > So there is a non-zero chance that we may point tapedrive at all three clusters at the same time, but only one would be considered primary (where tokens are earned). Does this remove the need for the config file?
> > > I guess not, but it might be worth thinking about what this file would look like with more than one source.
> > > @Nagaprasadvr what are your thoughts?
> > 
> > 
> > U mean simultaneously running tape node on multiple clusters ?
> > @zfedoran
> 
> Yeah, exactly this. The premise is that users can write wherever it is cheapest. Assuming they are okay with lower recovery guarantees.

What are differences will be in this case between a primary cluster ,(possibly mainnet ) and other clusters like devnet and testnet

Nagaprasadvr commented on cli/src/config.rs:
```diff
@@ -0,0 +1,148 @@
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct StorageConfig {
+    pub rocksdb_primary_path: String,
```
Yup exactly, but let's have Postgres instead of Sqllite

Nagaprasadvr commented on cli/src/config.rs:
```diff
@@ -0,0 +1,264 @@
+    pub fn load() -> anyhow::Result<Self> {
+        let home_dir = dirs::home_dir()
+            .ok_or_else(|| anyhow::anyhow!("Could not determine home directory"))?;
+        let config_path = home_dir.join("tape.toml");
```
lets load the config file path from cli args like cargo run -- --config ./tape.toml

Nagaprasadvr reviewed (changes requested):
looking good

zfedoran commented:
Lots of good stuff in here. 

@4rjunc if you'd like, I'd love to give you access to the dev channel in discord

4rjunc commented on cli/src/config.rs:
```diff
@@ -0,0 +1,264 @@
+    pub fn load_from_path<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+        let path = path.as_ref();
+        if !path.exists() {
+            return Err(anyhow::anyhow!("Configuration file not found"));
```
rn the tape.toml is created in home directory. so we don't need to add it to .gitignore

4rjunc commented:
thank you!

@zfedoran i haven't joined the discord yet. i will join now. my username: `4rjunc`

aarjn committed:
Squash

aarjn committed:
fix: removed duplicate lines

Nagaprasadvr reviewed (changes requested):
Few things to fix

Nagaprasadvr commented:
> This is looking really awesome @4rjunc. One more thing I'd love to see in here is the miner name. The current approach uses the word "default", which is fine, but we might as well put it in the config.
> 
> @Nagaprasadvr what are your thoughts?
> 
> (related issue: #36)

Yeah definitely ig i missed it but it's a good addition

aarjn committed:
fix: better names to structs and vars

aarjn committed:
feat: move config load-up -> Context:try_build

4rjunc commented:
> This is looking really awesome @4rjunc. One more thing I'd love to see in here is the miner name. The current approach uses the word "default", which is fine, but we might as well put it in the config.
> 
> @Nagaprasadvr what are your thoughts?
> 
> (related issue: #36)

thank you!

you mean to use `miner_address` from config in [this](https://github.com/spool-labs/tape/blob/690d05b949315799231eced0a8aade60f86a7873/cli/src/cli.rs#L96C9-L96C22) args place right  ?

Nagaprasadvr commented:
Hey @4rjunc ig we have changes from previous commits  did u rebase this ?

zfedoran commented:
> > This is looking really awesome @4rjunc. One more thing I'd love to see in here is the miner name. The current approach uses the word "default", which is fine, but we might as well put it in the config.
> > @Nagaprasadvr what are your thoughts?
> > (related issue: #36)
> 
> thank you!
> 
> you mean to use a custom name instead of "default" in this [line](https://github.com/spool-labs/tape/blob/690d05b949315799231eced0a8aade60f86a7873/cli/src/commands/network.rs#L139) right ?

Yes, exactly this

Nagaprasadvr commented on cli/src/config.rs:
```diff
@@ -0,0 +1,340 @@
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct MiningConfig {
+    pub num_cores: usize,
```
add miner_name here

Nagaprasadvr commented on cli/src/cli.rs:
```diff
@@ -246,12 +241,74 @@ pub struct Context {
-        let rpc_url = cli.cluster.rpc_url();
+        
+        // loading up configs
+        let config = match TapeConfig::load(&cli.config) {
```
dont match the error just return it use ?

4rjunc replied:
make sense now. it reduced the code. pushed the changes. please check. i will start working on remaining part tomorrow onwards

aarjn committed:
fix: add `miner_name`

aarjn committed:
feat: reduce match block of config loadup

Nagaprasadvr commented on tape.example.toml:
```diff
@@ -0,0 +1,27 @@
+[mining_config]
```
```suggestion
[mining]
```

Nagaprasadvr commented on cli/src/cli.rs:
```diff
@@ -246,12 +240,17 @@ pub struct Context {
+            commitment_level)
         );
-        let keypair_path = get_keypair_path(cli.keypair_path.clone());
+        let keypair_path = get_keypair_path(Some(PathBuf::from(&*shellexpand::tilde(&config.identity.keypair_path))));
```
Modify this fn to receive &config.identity.keypair_path as an arg

Nagaprasadvr commented on cli/src/config.rs:
```diff
@@ -0,0 +1,341 @@
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct IdentityConfig {
```
Let's move this IdentityConfig to SolanaConfig since we only have keypair_path we can just move that field to SolanaConfig

Nagaprasadvr commented on cli/src/config.rs:
```diff
@@ -0,0 +1,341 @@
+                match Self::load_from_path(&default_path) {
+                    Ok(config) => Ok(config),
+                    Err(TapeConfigError::ConfigFileNotFound) => {
+                        print_info("tape.toml not found, creating default configuration...");
```
Mention the path where its stored

Nagaprasadvr reviewed (changes requested):
Few nits

aarjn committed:
imp: change the args in `get_keypair_path` fn

aarjn committed:
imp: move the `keypair_path` -> `Solana`, remove `Identity`

aarjn committed:
imp: logs default config creation path

aarjn committed:
feat: add `config` field to `Context`

> rm keypair_path
> update keypair_path() helper function

aarjn committed:
feat: add config for mine

A

aarjn committed:
feat: add config for web

aarjn committed:
feat: add config for archive

aarjn committed:
feat: add config to snapshot

> open_read_only_store_conn

aarjn committed:
feat: add `max_transaction_retries`

aarjn committed:
feat: add `miner_name`

aarjn committed:
cleanup: delete unused code

Nagaprasadvr commented:
looks good

zfedoran commented:
closing, there are about 100k lines of diffs at this point in the proof of stake version, will review this once it is time for this

aarjn commented:
Okay, will open a new PR for this when tapedrive v2 is open
